### PR TITLE
src/main: add rauc mount command

### PR DESCRIPTION
**Pull request by jluebbe**
This allows easy inspection of bundle contents. As this uses squashfs in the
kernel (as `rauc install` does), this also doesn't allow --no-verify.

### CODE REVIEW

File: src/main.c [hunk 1]
```diff
@@ -1683,6 +1683,56 @@ static gboolean service_start(int argc, char **argv)
...
+	}
+
+	g_message("bundle mounted at: %s", bundle->mount_point);
+
+	/* The mount keeps the block device active. */
+	exit(0);
+
+out:
+	r_exit_status = res ? 0 : 1;
+	return TRUE;
```
Comment (Jarsop) [inline src/main.c:1733]: Why using `exit` and setting `r_exit_status` with a ternary. `out` label can it not exist that in case of error right ?

File: src/main.c [hunk 2]
```diff
@@ -1683,6 +1683,56 @@ static gboolean service_start(int argc, char **argv)
...
+
+	if (argc < 3) {
+		g_printerr("A file name must be provided\n");
+		r_exit_status = 1;
+		return FALSE;
+	}
+
+	if (argc > 3) {
+		g_printerr("Excess argument: %s\n", argv[3]);
+		r_exit_status = 1;
```
Comment (Jarsop) [inline src/main.c:1701]:
```suggestion
```
Redundant with `out` label logic

**Comment by Jarsop** [comment]
Perhaps is silly as question :stuck_out_tongue:  but why the `mount_start` can be return `TRUE` in some error case ?

File: src/main.c [hunk 3]
```diff
@@ -1683,6 +1683,56 @@ static gboolean service_start(int argc, char **argv)
...
+	res = mount_bundle(bundle, &error);
+	if (!res) {
+		g_printerr("%s\n", error->message);
+		g_clear_error(&error);
+		goto out;
+	}
+
+	g_message("bundle mounted at: %s", bundle->mount_point);
+
+	/* The mount keeps the block device active. */
```
Comment (zyga) [inline src/main.c:1728]: How is this comment related to the `exit` call?

**Review by zyga — COMMENTED** [review]
Quick look with one question inline.

File: src/main.c (hunk 3 above)
Comment (jluebbe) [reply src/main.c:1728]: The device mapper target and loopback devices are configured to remove themselves on close. So we don't need to handle cleanup in RAUC, a simple unmount on the is enough. I'll extend the comment accordingly.

File: docs/reference.rst [hunk 4]
```diff
@@ -670,6 +670,7 @@ Command Line Tool
     extract               Extract the bundle content
     install               Install a bundle
     info                  Show file information
+    mount                 Mount RAUC bundle
```
Comment (Bastian-Krause) [inline docs/reference.rst:673]:
```suggestion
    mount                 Mount a bundle
```
This would be consistent with the other command's descriptions.

File: src/main.c (hunk 2 above)
Comment (Bastian-Krause) [reply src/main.c:1701]: True. Why go to "out" here, but return early above?

File: src/main.c [hunk 5]
```diff
@@ -1844,6 +1895,7 @@ static void cmdline_handler(int argc, char **argv)
 #if ENABLE_SERVICE == 1
 		{SERVICE, "service", "service", "Start RAUC service", service_start, service_group, TRUE},
 #endif
+		{MOUNT, "mount", "mount <BUNDLENAME>", "Mount RAUC bundle", mount_start, NULL, TRUE},
```
Comment (Bastian-Krause) [inline src/main.c:1898]:
```suggestion
		{MOUNT, "mount", "mount <BUNDLENAME>", "Mount a bundle", mount_start, NULL, TRUE},
```

File: src/main.c [hunk 6]
```diff
@@ -1864,6 +1916,7 @@ static void cmdline_handler(int argc, char **argv)
 			"  extract\t\tExtract the bundle content\n"
 			"  install\t\tInstall a bundle\n"
 			"  info\t\t\tShow file information\n"
+			"  mount\t\t\tMount RAUC bundle\n"
```
Comment (Bastian-Krause) [inline src/main.c:1919]:
```suggestion
			"  mount\t\t\tMount a bundle\n"
```

**Review by Bastian-Krause — CHANGES_REQUESTED** [review]

File: src/main.c (hunk 2 above)
Comment (jluebbe) [reply src/main.c:1701]: Yes, that's inconsistent, will fix.

File: src/main.c (hunk 6 above)
Comment (jluebbe) [reply src/main.c]: Changed.

File: src/main.c (hunk 5 above)
Comment (jluebbe) [reply src/main.c]: Changed.

File: docs/reference.rst (hunk 4 above)
Comment (jluebbe) [reply docs/reference.rst]: Changed.

File: src/main.c (hunk 3 above)
Comment (jluebbe) [reply src/main.c]: I've expanded the comment.

File: src/main.c (hunk 1 above)
Comment (jluebbe) [reply src/main.c]: I've made this consistent. `r_exit_status` is now set to `1` in `out:` and we return `FALSE`.

**Comment by jluebbe** [comment]
Rebased.

**Review by Bastian-Krause — DISMISSED** [review]
Looks good to me.

**Comment by Jarsop** [comment]
Same for me

**Comment by ejoerns** [comment]
In general, I like the idea of the command.

However, some more or less notable things that I would like to mention are:

a) Maybe we should leave a note in the help that this is not a "normal operation" command as I fear that people would tend to see this as something that is totally fine with the flow of RAUC and will work around the standard install strategies even more. For those who really know what they do and why, this is fine, but I would not unnecessarily encourage people to abuse 'rauc mount'.

Thus possibly something like

``` patch
--- a/docs/reference.rst
+++ b/docs/reference.rst
@@ -670,6 +670,7 @@ Command Line Tool
      info                  Show file information
-+    mount                 Mount a bundle
++    mount                 Mount a bundle for debugging and inspection purpose
      service               Start RAUC service
```

What do you think?


b) The command requires root privileges as far as I see, thus it is possibly primarily useful for the target or should be used on the host too?
It also states to be similarly to `rauc install` but ignores the system.conf file we have on the target. Thus we must provide the required arguments (keyring, mount prefix, etc.) by hand. Especially the `--mount` argument appears to be a bit unexpected/inconvenient to me (We omit the destination argument a normal mount call has, but also do not use the configured one. Thus end up in the rauc default mount path..).
The reason for doing it only this way is obviously the need to introduce a new D-Bus method otherwise. Thus it might be a compromise we need to accept, but I would like to make sure that we have thought of the use cases properly before.

c) We do not provide an 'unmount' command. Probably the command should print the instruction to unmount the bundle to shell? Like

    Mounted bundle at /mnt/rauc/bundle. Type 'umount /mnt/rauc/bundle' to unmount.

d) The command is meant to be a debugging tool by default (if I get it right). Thus I would propose to drop a note in `Debugging RAUC` section in docs/using.rst.

e) We should take more care about updating the man page (note to myself, too) when extending the CLI. Here the added command should be reflected, too.

That's all I guess ;)

**Comment by jluebbe** [comment]
> In general, I like the idea of the command.
> 
> However, some more or less notable things that I would like to mention are:
> 
> a) Maybe we should leave a note in the help that this is not a "normal operation" command as I fear that people would tend to see this as something that is totally fine with the flow of RAUC and will work around the standard install strategies even more. For those who really know what they do and why, this is fine, but I would not unnecessarily encourage people to abuse 'rauc mount'.
> 
> Thus possibly something like
> 
> ```diff
> --- a/docs/reference.rst
> +++ b/docs/reference.rst
> @@ -670,6 +670,7 @@ Command Line Tool
>       info                  Show file information
> -+    mount                 Mount a bundle
> ++    mount                 Mount a bundle for debugging and inspection purpose
>       service               Start RAUC service
> ```
> 
> What do you think?

Good idea.

> b) The command requires root privileges as far as I see, thus it is possibly primarily useful for the target or should be used on the host too?

Yes, this is mainly useful on the target.

> It also states to be similarly to `rauc install` but ignores the system.conf file we have on the target. Thus we must provide the required arguments (keyring, mount prefix, etc.) by hand. Especially the `--mount` argument appears to be a bit unexpected/inconvenient to me (We omit the destination argument a normal mount call has, but also do not use the configured one. Thus end up in the rauc default mount path..).

Hmm, it should use the normal system.conf, just like rauc info or install. If this this doesn't work for you, we need to check why. Would you still prefer to have an (optional) mount point argument after the bundle?

> The reason for doing it only this way is obviously the need to introduce a new D-Bus method otherwise. Thus it might be a compromise we need to accept, but I would like to make sure that we have thought of the use cases properly before.

A mount created via the service wouldn't be visible due to the configuration of mount namespaces. Also, it would need need more coordination on unmount. As it's intended to be a development command, a DBus interface wouldn't be useful.

> c) We do not provide an 'unmount' command. Probably the command should print the instruction to unmount the bundle to shell? Like
> 
> ```
> Mounted bundle at /mnt/rauc/bundle. Type 'umount /mnt/rauc/bundle' to unmount.
> ```

Yes.

> d) The command is meant to be a debugging tool by default (if I get it right). Thus I would propose to drop a note in `Debugging RAUC` section in docs/using.rst.

Yes.

> e) We should take more care about updating the man page (note to myself, too) when extending the CLI. Here the added command should be reflected, too.

Right.

> That's all I guess ;)

**Comment by ejoerns** [comment]
> > It also states to be similarly to `rauc install` but ignores the system.conf file we have on the target. Thus we must provide the required arguments (keyring, mount prefix, etc.) by hand. Especially the `--mount` argument appears to be a bit unexpected/inconvenient to me (We omit the destination argument a normal mount call has, but also do not use the configured one. Thus end up in the rauc default mount path..).
> 
> Hmm, it should use the normal system.conf, just like rauc info or install. If this this doesn't work for you, we need to check why. Would you still prefer to have an (optional) mount point argument after the bundle?

Yes and no ;)

If you provide `-c /path/to/config` or if system.conf is located at the default path, it will use it.

For `rauc info` the situation is similar. For `rauc install`, this is not necessary of course as the service executes the actual command.

We hit the host vs target tool problematic here again I guess. From my point of view, a 'host' tool should not use system.conf while a target tool should use the system.conf via service only (for D-Bus case of course). Thus, probably an (at least optional) mount point argument would be a proper solution. Then it can be called with --keyring argument (which is similar to other host tools). The fact that this is a 'host' tool that can primarily be used on the target is probably something we have to endure ;)

**Comment by jluebbe** [comment]
This was just a straight rebase to make the following fixes easier to review.

File: src/main.c [hunk 7]
```diff
@@ -1683,6 +1683,59 @@ static gboolean service_start(int argc, char **argv)
...
+		goto out;
+	}
+
+	if (argc > 3) {
+		g_printerr("Excess argument: %s\n", argv[3]);
+		goto out;
+	}
+
+	bundlelocation = resolve_bundle_path(argv[2]);
+	if (bundlelocation == NULL)
```
Comment (zyga) [inline src/main.c:1704]: Is there some error that can be printed in this case?

**Review by zyga — COMMENTED** [review]
Looks good, just one small question inline.

File: test/rauc.t [hunk 8]
```diff
@@ -307,6 +307,19 @@ test_expect_success PKCS11 "rauc bundle with PKCS11 (key mismatch)" "
...
 "
 
+test_expect_success ROOT "rauc mount" "
+  test ! -f /mnt/rauc/bundle/manifest.raucm &&
+  cp -L ${SHARNESS_TEST_DIRECTORY}/good-bundle.raucb ${TEST_TMPDIR}/ &&
+  test_when_finished rm -f ${TEST_TMPDIR}/good-bundle.raucb &&
+  test_when_finished umount /mnt/rauc/bundle &&
+  ls ${TEST_TMPDIR} &&
+  rauc --keyring $SHARNESS_TEST_DIRECTORY/openssl-ca/dev-ca.pem \
+    mount ${TEST_TMPDIR}/good-bundle.raucb &&
```
Comment (zyga) [inline test/rauc.t:317]: Do you need to unmount this or is the test scaffolding responsible for that?

File: src/main.c (hunk 7 above)
Comment (jluebbe) [reply src/main.c:1704]: ``resolve_bundle_path()`` prints an error. I've added a comment to the ``goto out;`` line.

File: test/rauc.t (hunk 8 above)
Comment (jluebbe) [reply test/rauc.t:317]: That is handled by the ``test_when_finished umount /mnt/rauc/bundle`` above.

**Comment by jluebbe** [comment]
I've applied most of the changes requested by @ejoerns, except changing the configuration handling. That should be done as a separate PR and ensure that the on-target commands check that the config is loaded correctly (to avoid confusion when a different config location is used).

**Review by ejoerns — APPROVED** [review]
Changes look fine for me, thx!
